pkg/operator/controllers/imageconfig/image_controller.go

```diff
@@ -0,0 +1,154 @@
+package imageconfig
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/Azure/go-autorest/autorest/azure"
+	configv1 "github.com/openshift/api/config/v1"
+	configclient "github.com/openshift/client-go/config/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	aroclient "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned"
+	"github.com/Azure/ARO-RP/pkg/operator/controllers"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient"
+)
+
+const imageConfigResource = "cluster"
+
+type Reconciler struct {
+	log       *logrus.Entry
+	arocli    aroclient.Interface
+	configcli configclient.Interface
+}
+
+func NewReconciler(log *logrus.Entry, arocli aroclient.Interface, configcli configclient.Interface) *Reconciler {
+	return &Reconciler{
+		log:       log,
+		arocli:    arocli,
+		configcli: configcli,
+	}
+}
+
+// watches the ARO object for changes and reconciles image.config.openshift.io/cluster object.
+// - If blockedRegistries is not nil, makes sure required registries are not added
+// - If AllowedRegistries is not nil, makes sure required registries are added
+// - Fails fast if both are not nil, unsupported
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	// Get cluster
+	instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, request.Name, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Feature flag
+	if !instance.Spec.Features.ReconcileImageConfig {
+		return reconcile.Result{}, nil
+	}
+
+	// Check for cloud type
+	requiredRegistries, err := getCloudAwareRegistries(instance)
+	if err != nil {
+		// Not returning error as it will requeue again
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	// Get image.config yaml
+	imageconfig, err := r.configcli.ConfigV1().Images().Get(ctx, request.Name, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Fail fast if both are not nil
+	if imageconfig.Spec.RegistrySources.AllowedRegistries != nil && imageconfig.Spec.RegistrySources.BlockedRegistries != nil {
```

Comment:
This seems like a valid use case to me - is it unsupported?

Reply:
Yes, this is unsupported by upstream imageconfig. We can only have one or the other. Check in notes here https://docs.openshift.com/container-platform/4.6/openshift_images/image-configuration.html#images-configuration-allowed_image-configuration